mbedtls/src/ssl/context.rs

```diff
@@ -500,6 +511,250 @@ impl HandshakeContext {
     }
 }
 
+#[cfg(all(feature = "std", feature = "async"))]
+struct HandshakeFuture<'a, T>(&'a mut Context<AsyncIoAdapter<T>>);
+
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T> Future for HandshakeFuture<'_, T> {
+    type Output = Result<()>;
+    fn poll(mut self: Pin<&mut Self>, ctx: &mut TaskContext) -> std::task::Poll<Self::Output> {
+        self.0
+            .io_mut()
+            .ok_or(Error::NetInvalidContext)?
+            .ecx
+            .set(ctx);
+
+        let result = match self.0.handshake() {
+            Err(Error::SslWantRead) | Err(Error::SslWantWrite) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(e)),
+            Ok(()) => Poll::Ready(Ok(())),
+        };
+
+        self.0.io_mut().map(|v| v.ecx.clear());
+
+        result
+    }
+}
+
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T: Unpin + 'static> Context<AsyncIoAdapter<T>> {
+    pub async fn accept_async<IoType>(
+        config: Arc<Config>,
+        io: T,
+        hostname: Option<&str>,
+    ) -> IoResult<Context<AsyncIoAdapter<T>>>
+    where
+        AsyncIoAdapter<T>: IoCallbackUnsafe<IoType>,
+    {
+        let mut context = Self::new(config);
+        context
+            .establish_async(io, hostname)
+            .await
+            .map_err(|e| crate::private::error_to_io_error(e))?;
+        Ok(context)
+    }
+
+    pub async fn establish_async<IoType>(&mut self, io: T, hostname: Option<&str>) -> Result<()>
+    where
+        AsyncIoAdapter<T>: IoCallbackUnsafe<IoType>,
+    {
+        unsafe {
+            let mut io = Box::new(AsyncIoAdapter::new(io));
+
+            ssl_session_reset(self.into()).into_result()?;
+            self.set_hostname(hostname)?;
+
+            let ptr = &mut *io as *mut _ as *mut c_void;
+            ssl_set_bio(
+                self.into(),
+                ptr,
+                Some(AsyncIoAdapter::<T>::call_send),
+                Some(AsyncIoAdapter::<T>::call_recv),
+                None,
+            );
+
+            self.io = Some(io);
+            self.inner.reset_handshake();
+        }
+
+        HandshakeFuture(self).await
+    }
+}
+
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T: AsyncRead> AsyncRead for Context<AsyncIoAdapter<T>>
```

Comment:
Please move to async_io.rs (like in io.rs). Same for write.

Reply:
moved